openedx/features/qverse_features/registration/signals.py

```diff
@@ -1,20 +1,172 @@
+            output_file_rows[-1].update({'status': ''})
+            # Creating/Updating Edx User, Edx User Profile and QVerse User Profile
+            # will be an atomic operation. If one operation fails the previous
+            # successfull operations will be reverted and no changes will be applied.
+            with transaction.atomic():
+                edx_user, _ = _create_or_update_edx_user(row)
+                _create_or_update_edx_user_profile(edx_user)
+                _create_or_update_qverse_user_profile(edx_user, row, output_file_rows)
+        except (IntegrityError, ValueError):
+            LOGGER.exception('Error while creating/updating {} user and its profiles.'.format(row.get('regno')))
+            output_file_rows[-1].update({'status': 'Failed'})
+        except Department.DoesNotExist:
+            LOGGER.exception('The Department with id {} does not exist.'.format(row.get('departmentid')))
+            output_file_rows[-1].update({'status': 'Failed'})
+    csv_file.close()
+    _write_status(instance.admission_file.path, output_file_rows)
+
+
+def _create_or_update_edx_user(user_info):
+    """
+    Creates/Updates edx user from given information.
+
+    Arguments:
+        user_info - A dict containing user information
+
+    Returns:
+        edx_user - (User) A newly created/updated User object
+        created - (bool) A flag indicating that user is created or not
+    """
+    user_data = {
+                    'password': make_password(user_info.get('regno')),
+                    'email': user_info.get('email'),
+                    'first_name': user_info.get('firstname'),
+                    'last_name': user_info.get('surname'),
+                    'is_active': True
+                }
+    edx_user, created = User.objects.get_or_create(username=user_info.get('regno'), defaults=user_data)
+    if created:
+        LOGGER.info('{} user has been created.'.format(edx_user.username))
+    else:
+        edx_user.first_name = user_info.get('firstname')
+        edx_user.last_name = user_info.get('surname')
+        edx_user.email = user_info.get('email')
+        edx_user.save()
+        LOGGER.info('{} user has been updated.'.format(edx_user.username))
+    return edx_user, created
+
+
+def _create_or_update_edx_user_profile(edx_user):
+    """
+    Creates/Updates edx user profile.
+
+    Arguments:
+        edx_user - (User) An edx user instance whose profile is going to be created/updated
+
+    Returns:
+        created - (bool) A flag telling that user profile is created or not
+    """
+    full_name = '{} {}'.format(edx_user.first_name, edx_user.last_name)
+    _, created = UserProfile.objects.update_or_create(user=edx_user, defaults={'name': full_name})
+    if created:
+        LOGGER.info('{} edx profile has been created.'.format(edx_user.username))
+    else:
+        LOGGER.info('{} edx profile has been updated.'.format(edx_user.username))
+    return created
```

Comment:
You are not consuming this returned value in your original function. Why returning it from here?

Reply:
done.